cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java

```diff
@@ -0,0 +1,195 @@
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          delegate.onFinish(BodyConsumerAdapter.this.responder);
+        }
+      });
+      BodyConsumerAdapter.this.responder.execute();
+    } catch (Throwable t) {
+      onError(t, this.responder);
+    }
+  }
+
+  @Override
+  public void handleError(final Throwable cause) {
+    // When this method is called from netty-http, the response has already been sent, hence uses a no-op
+    // DelayedHttpServiceResponder for the onError call.
+    onError(cause, new DelayedHttpServiceResponder(responder) {
+      @Override
+      protected void doSend(int status, ByteBuffer content, String contentType,
+                            @Nullable Multimap<String, String> headers, boolean copy) {
+        // no-op
+      }
+
+      @Override
+      public void setTransactionFailureResponse(Throwable t) {
+        // no-op
+      }
+
+      @Override
+      public void execute() {
+        // no-op
+      }
+    });
+  }
+
+  /**
+   * Calls the {@link HttpContentConsumer#onError(HttpServiceResponder, Throwable)} method from a transaction.
+   */
+  private void onError(Throwable cause, final DelayedHttpServiceResponder responder) {
+    final Throwable failureCause = (cause instanceof WrappedException) ? cause.getCause() : cause;
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          delegate.onError(responder, failureCause);
+        }
+      });
+    } catch (Throwable t) {
+      responder.setTransactionFailureResponse(t);
+    } finally {
+      responder.execute();
+    }
+  }
+
+  private Transactional createTransactional(final TransactionContext txContext,
+                                            final DatasetContext datasetContext) {
+    return new Transactional() {
+      @Override
+      public void execute(final TxRunnable runnable) {
+        try {
+          txContext.start();
+          try {
+            ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(contextClassLoader);
+            try {
+              runnable.run(datasetContext);
+            } finally {
+              ClassLoaders.setContextClassLoader(oldClassLoader);
+            }
+            txContext.finish();
+          } catch (Throwable t) {
+            txContext.abort(new TransactionFailureException("User handler exception", t));
+          }
+        } catch (Throwable t) {
+          throw new WrappedException(t);
+        }
+      }
+    };
+  }
+
+  /**
+   * A {@link RuntimeException} for wrapping any {@link Throwable}. It is needed because {@link Transactional}
+   * only throws {@link RuntimeException} so that in {@link #onError(Throwable, DelayedHttpServiceResponder)},
+   * we can unwrap and get the actual cause.
+   */
+  private static final class WrappedException extends RuntimeException {
```

Comment:
This is like ExecutionException? Why not keep the TransactionFailureException hierarchy?